mindmeld/components/dialogue.py

```diff
@@ -957,7 +977,25 @@ def __call__(self, request, responder):
                 self._prompt_turn = True
 
         # Finish slot-filling and return to handler
-        return self._end_slot_fill(request, responder)
+        return self._end_slot_fill(request, responder, self._app.async_mode)
+
+    def call_sync(self, request, responder):
+        """The slot-filling call for synchronous apps
+
+        Args:
+            request (Request): The request object.
+            responder (DialogueResponder): The responder object.
+        """
+        self.__call__(request, responder)
+
+    async def call_async(self, request, responder):
+        """The slot-filling call for asynchronous apps
+
+        Args:
+            request (Request): The request object.
+            responder (DialogueResponder): The responder object.
+        """
+        self.__call__(request, responder)
```

Comment:
Do we have to do `await self.__call__(request, responder)` here?

Reply:
You only `await` coroutines, `__call__` is a regular function. An `async` _allows_ you to `await` coroutines and to be `await`ed by something further up the stack. It does not require you to only invoke coroutines, you're free to call synchronous functions however you like.

You also don't generally want to call dunder methods directly because they're just the implementation for the low-level interface. In this case, `__call__` makes instances of this class callable so you can just do that:

```python
async def call_async(self, request, responder):
    self(request, responder)
```

Reply:
Expanding on this a bit more... there's really no reason to have this method or `call_sync`. `__call__` isn't asynchronous and can't be made asynchronous so you might as well drop both `call_sync` and `call_async` and just invoke instances of this class as callables.

```python
filler = AutoEntityFilling()
filler(my_request, my_responder)
```

Reply:
I think the reason I did the two functions was that the application file is where these were being called and passed into the ``add_dialogue_rule`` method, which requires an asynchronous handler to be passed when app is in async mode. Having a common __call__ doesn't allow for that? What's the best alternative then?

Reply:
That makes sense, I'm with you now. Carry on!